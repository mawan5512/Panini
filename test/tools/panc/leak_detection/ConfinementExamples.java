/*
 * This file is part of the Panini project at Iowa State University.
 *
 * The contents of this file are subject to the Mozilla Public License
 * Version 1.1 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/.
 *
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
 * for the specific language governing rights and limitations under the
 * License.
 *
 * For more details and the latest version of this code please see
 * http://paninij.org
 *
 * Contributor(s): Yuheng Long
 */

/*
 * @test
 * @summary Ensure Capsule state will not be leaked to other Capsule.
 * @compile/ref=ConfinementExamples.out -XDrawDiagnostics  ConfinementExamples.java
 */

class TestC {
	TestC next;
	void setNext(TestC next) { this.next = next; }
}

capsule C {
	void test(TestC tc) {  }
}

capsule M (C c) {
	TestC tc = new TestC();
	void mtest() {
		tc.setNext(tc);
		c.test(tc);
	}

	TestC mtest2() {
		return tc;
	}
}

capsule ConfinementExamples {
	design {
		C c; M m;
		m(c);
	}

	void run() {
		m.mtest();
		m.mtest2();
	}
}
