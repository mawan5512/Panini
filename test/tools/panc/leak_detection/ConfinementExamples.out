ConfinementExamples.java:39:24: compiler.warn.confinement.violation: tc, M, mtest
ConfinementExamples.java:43:24: compiler.warn.confinement.violation: tc, M, mtest2
2 warnings
